src/processor/stackwalker.cc

```diff
@@ -164,30 +175,42 @@ bool Stackwalker::Walk(
     }
 
     // Keep track of the number of dubious frames so far.
-    switch (frame.get()->trust) {
-       case StackFrame::FRAME_TRUST_NONE:
-       case StackFrame::FRAME_TRUST_SCAN:
-       case StackFrame::FRAME_TRUST_CFI_SCAN:
-         scanned_frames++;
-         break;
-      default:
-        break;
+    if (!is_trusted_frame(frame.get())) {
+      scanned_frames++;
     }
 
+    // XXX(swatinem): We speculatively walk over NULL/near-NULL frames,
+    // using the same condition as `TerminateWalk`.
+    bool is_speculative_frame = frame.get()->ReturnAddress() < (1 << 12);
+
     // Add the frame to the call stack.  Relinquish the ownership claim
     // over the frame, because the stack now owns it.
     stack->frames_.push_back(frame.release());
     if (stack->frames_.size() > max_frames_) {
       // Only emit an error message in the case where the limit
       // reached is the default limit, not set by the user.
-      if (!max_frames_set_)
+      if (!max_frames_set_) {
         BPLOG(ERROR) << "The stack is over " << max_frames_ << " frames.";
+        if (is_speculative_frame) {
+          delete stack->frames_.back();
+          stack->frames_.pop_back();
+        }
+      }
       break;
     }
 
     // Get the next frame and take ownership.
     bool stack_scan_allowed = scanned_frames < max_frames_scanned_;
-    frame.reset(GetCallerFrame(stack, stack_scan_allowed));
+    StackFrame* next_frame = GetCallerFrame(stack, stack_scan_allowed);
+    if (is_speculative_frame && (!next_frame || !is_trusted_frame(next_frame))) {
```

Comment:
Without base pointers, which happens when FPO is on and the platform ABI does not mandate base pointers, then the next frame will inevitably be scanned. However, that's fine as long as the frame is not again "speculative". 

Would this change still work when doing this:

```c
    if (is_speculative_frame(frame) && (!next_frame || !is_speculative_frame(next_frame))) {
```